### PR TITLE
Incident per feed widget

### PR DESCRIPTION
**Pull request by ShahafBenYakir**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/20180

## Description
New widget that displays the number of incidents per feed.

## Screenshots
Cortex XSOAR - Dashboards 2020-03-16 12-15-48

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
   - [x] No

### CODE REVIEW

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    demisto.results({
+        'total': len(data),
+        'data': [{
+            'Feed Name': key,
+            'Number Of Incidents': val
+        } for key, val in data.items()]})
+    demisto.results('Done')
+
+
+if __name__ == "__main__":
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]:
```suggestion
if __name__ in ('__main__', '__builtin__', 'builtins'):
```

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    default_from_date = get_default_from_date('30 days')
+    from_date = demisto.args().get('from', default_from_date)
+    data = get_incidents_per_feed(from_date)
+    demisto.results({
+        'total': len(data),
+        'data': [{
+            'Feed Name': key,
+            'Number Of Incidents': val
+        } for key, val in data.items()]})
+    demisto.results('Done')
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]: remove this

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    contents = demisto.executeCommand("getIncidents", {"fromdate": from_date})[0]['Contents']
+    incidents = contents['data']
+    size = len(incidents)
+    total = contents['total']
+    page = 1
+
+    while total > size:
+        contents = demisto.executeCommand("getIncidents", {"fromdate": from_date, "page": page})[0]['Contents']
+        new_incidents = contents['data']
+        incidents = incidents + new_incidents
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]:
```suggestion
        incidents += new_incidents
```

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        size = len(incidents)
+        page = page + 1
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]:
```suggestion
        page += 1
```

**Comment by hod-alpert** [comment]
Code itself looks great.

Two questions:

1. What about the False Positive incidents and the ration between them?
2. How long does it take to load? As I recall I had a solution pretty much like this, but it was not scalable and it took about half a minute.

Here you do O(incidents) queries, this could be tens of thousends of queries and this could take minutes to load, this was the reason @anara123 said he'd rather wait for a better solution then presenting one that gives bad experience.

Anyways- @aviadl showed a bit more efficient way to do it by querying indicators first and collecting thier InvestigationIDs and the doing big queries for incidents with these incidents ids saparated by "or".

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed_test.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+def execute_command(name, args=None):
+    if name == 'getIncidents':
+        return INCIDENTS_TO_RETURN
+    else:
+        return None
+
+
+def search_indicators(query):
+    # query = args.get('indicator_query')
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed_test.py]: remove comment

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
+	"shouldCommit": false,
 	"fromDate": "0001-01-01T00:00:00Z",
 	"toDate": "0001-01-01T00:00:00Z",
 	"period": {
 		"by": "",
 		"byTo": "",
 		"byFrom": "days",
 		"toValue": null,
-		"fromValue": 180,
+		"fromValue": 30,
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Was downloaded like this from Demisto.

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 7]
```diff
@@ -1,15 +1,19 @@
 {
 	"id": "Threat Intelligence Management",
 	"version": -1,
-	"modified": "2020-02-23T17:14:04.927422326Z",
+	"modified": "2020-03-16T09:58:47.497959016Z",
+	"sortValues": null,
+	"vcShouldIgnore": false,
+	"commitMessage": "",
+	"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove these

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 8]
```diff
@@ -28,6 +32,10 @@
 				"id": "14e61cc5-3584-43d4-893b-16f9db4ef87f",
 				"version": 1,
 				"modified": "2020-01-14T13:20:34.180114532Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 9]
```diff
@@ -133,11 +141,15 @@
...
-			"h": 2,
+			"h": 1,
 			"widget": {
 				"id": "1c789a4b-d000-4439-85bb-b113dafbfac9",
 				"version": 1,
 				"modified": "2020-01-14T13:23:27.49284693Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 10]
```diff
@@ -236,14 +248,18 @@
...
-			"h": 2,
+			"h": 1,
 			"widget": {
 				"id": "530fa858-2e59-4ef9-8b44-f47db1011b36",
 				"version": 1,
 				"modified": "2020-01-14T13:24:57.840603414Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 11]
```diff
@@ -341,15 +357,19 @@
...
+			"w": 3,
 			"h": 2,
 			"widget": {
 				"id": "e0aded97-34a1-45f3-8b5b-a54b43c13f5f",
 				"version": 1,
 				"modified": "2020-01-14T13:30:13.387314353Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 12]
```diff
@@ -459,6 +479,10 @@
 				"id": "8dd2c47e-84ec-4309-8ec7-8216c88aa004",
 				"version": 1,
 				"modified": "2020-01-14T13:32:18.782752002Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 13]
```diff
@@ -560,15 +584,19 @@
...
+			"w": 3,
 			"h": 2,
 			"widget": {
 				"id": "0aaa7c20-668c-4cda-86d5-8e0744f9db1d",
 				"version": 1,
 				"modified": "2020-01-14T14:02:22.181828299Z",
+				"sortValues": null,
+				"vcShouldIgnore": false,
+				"commitMessage": "",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 14]
```diff
@@ -665,7 +693,69 @@
...
+			"x": 6,
+			"y": 2,
+			"i": "34a8d5e0-66e7-11ea-9204-9b2092cbe074",
+			"w": 6,
+			"h": 2,
+			"widget": {
+				"id": "incidentPerFeed",
+				"version": 7,
+				"modified": "2020-03-15T18:03:00.066866266Z",
+				"sortValues": null,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 15]
```diff
@@ -665,7 +693,69 @@
...
+				"propagationLabels": [
+					"all"
+				],
+				"vcShouldIgnore": false,
+				"commitMessage": "Widget imported",
+				"shouldCommit": false,
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Remove

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 16]
```diff
@@ -665,7 +693,69 @@
...
+			}
 		}
 	],
-	"fromVersion": "5.5.0"
-}
+	"system": false,
+	"isCommon": false,
+	"fromVersion": "5.5.0",
+	"shared": true,
+	"locked": false
```
Comment (hod-alpert) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]: Leave only the fromVersion

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.yml [hunk 17]
```diff
@@ -0,0 +1,15 @@
+comment: Summs the number of incident per OOB feed.
+commonfields:
+  id: ThreatIntelligenceManagementGetIncidentsPerFeed
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.yml]:
```suggestion
  id: ThreatIntelManagementGetIncidentsPerFeed
```

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.yml [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+  version: -1
+enabled: false
+name: ThreatIntelligenceManagementGetIncidentsPerFeed
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.yml]:
```suggestion
name: ThreatIntelManagementGetIncidentsPerFeed
```

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 19]
```diff
@@ -1,20 +1,19 @@
...
 		"by": "",
 		"byTo": "",
 		"byFrom": "days",
 		"toValue": null,
-		"fromValue": 180,
+		"fromValue": 30,
 		"field": ""
 	},
 	"fromDateLicense": "0001-01-01T00:00:00Z",
 	"name": "Threat Intelligence Management",
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]:
```suggestion
	"name": "Threat Intel Management",
```

File: Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json [hunk 20]
```diff
@@ -665,7 +658,61 @@
...
+				"id": "incidentPerFeed",
+				"version": 7,
+				"modified": "2020-03-15T18:03:00.066866266Z",
+				"propagationLabels": [
+					"all"
+				],
+				"name": "Incidents Per Feed",
+				"dataType": "scripts",
+				"widgetType": "table",
+				"query": "ThreatIntelligenceManagementGetIncidentsPerFeed",
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Dashboards/dashboard-Threat_Intelligence_Management.json]:
```suggestion
				"query": "ThreatIntelManagementGetIncidentsPerFeed",
```

File: Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    Query all incidents starting a given date.
+    :param from_date: string
+        The date to query incidents from.
+    :return: list
+        List of incidents.
+    """
+    command_res = demisto.executeCommand('getIncidents', {'fromdate': from_date})
+    if is_error(command_res):
+        LOG(f'Error executing "getIncidents fromdate: {from_date}"')
```
Comment (Itay4) [inline Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]: why not just `return_error(command_res)` ?

Comment (ShahafBenYakir) [reply Packs/ThreatIntelligenceManagement/Scripts/ThreatIntelligenceManagementGetIncidentsPerFeed/ThreatIntelligenceManagementGetIncidentsPerFeed.py]: Definitely forgot it also does LOG.

**Comment by ShahafBenYakir** [comment]
@Itay4 This is not really possible (the machine dies right after it successfully installing content and running the test PB)
Will create an intentional failure that will keep the machine alive.

**Comment by kirbles19** [comment]
@ShahafBenYakir Done.

**Pull request merged**